### ISSUE TITLE
Expose a way for consumers to modify query results before rendering the specification

### ISSUE THREAD

**Issue by benstoltz** [Esri/cedar#229]
Post Cedar.update, but prior to Cedar._renderspec we should introduce a mutateData style function.

**Comment by tomwayson** [comment]
Actually this needs to happen w/in `update`, post creating of spec, but before `_renderSpec()` is called.

**Comment by tomwayson** [comment]
One idea

Add an optional `onQuery` property to `dataset` that is a callback function that, if present, will be called after the dataset is queried and passed the array of features returned by the query giving the consumer the opportunity to modify those features before they are assigned to the spec values and rendered. 

Example: 

```js
chart.dataset = {
  "url":"https://services.arcgis.com/uDTUpUPbk8X8mXwl/arcgis/rest/services/Public_Schools_in_Onondaga_County/FeatureServer/0",
  "onQuery": function(features) {
    return features.map(function(feature) {
      // TODO: modify feature attributes here
      return feature;
    });
  },
  "mappings":{
    "group": {"field":"District","label":"District"},
    "count": {"field":"Number_of","label":"Student Number_of"}
  }
}
```

**Comment by tomwayson** [comment]
Another idea would be to break `update()` into into individual functions each part of the public API. Here's what I'm thinking (names not final):
1. `compileSpecification(values)` - roughly [these lines](https://github.com/Esri/cedar/blob/82aac15187dbd1fdc8de18bad3116c6926bb2b6e/src/cedar.js#L403-L420) though I'd suggest adding an optional values argument that would set `spec.data[0].values` (see below)
1. `query(callback)` - roughly [these lines](https://github.com/Esri/cedar/blob/82aac15187dbd1fdc8de18bad3116c6926bb2b6e/src/cedar.js#L430-L453), however, those depend on [these](https://github.com/Esri/cedar/blob/82aac15187dbd1fdc8de18bad3116c6926bb2b6e/src/cedar.js#L405-L406), which are a total mystery to me
1. `renderSpecification(spec, callback)` - effectively just make `_renderSpecification()` part of public API - I 

So the consumer would do something like:

```js
chart.query((features) => {
  // TODO: modify the features here as needed
  let spec = chart.compileSpecification(features); // this sets spec.data[0].values = features
  chart.renderSpecification(spec, clb); // where clb is a callback defined somewhere above
});
```

May need to account for some of [these lines](https://github.com/Esri/cedar/blob/82aac15187dbd1fdc8de18bad3116c6926bb2b6e/src/cedar.js#L387-L400) in those functions too, not sure.

I think this approach gives more flexibility than the `onQuery` callback above, and I think `renderSpecification()` could address #213 (at least partially) as well. However, I think it will be harder to implement and require more tests to ensure no regressions are made. Also, these new public API functions will be hard to get rid of, so we may need to think about how they would look/work in a multi-dataset, multi-query, and/or multi-series version of Cedar and design accordingly.

cc @benstoltz @ajturner

**Comment by tomwayson** [comment]
Under no circumstances should any function in this library or any other be called "mutateData"

**Comment by dbouwman** [comment]
@tomwayson 

## fn()'s in Json no bueno
I'm sure you likely know this, but you can't have functions in json... what I mean is that even the GH json view shows the json below as invalid b/c there is a function in it.

```json
{  "url":"https://services.arcgis.com/uDTUpUPbk8X8mXwl/arcgis/rest/services/Public_Schools_in_Onondaga_County/FeatureServer/0",
  "onQuery": function(features) {
    return features.map(function(feature) {
      // TODO: modify feature attributes here
      return feature;
    });
  },
  "mappings":{
    "group": {"field":"District","label":"District"},
    "count": {"field":"Number_of","label":"Student Number_of"}
  }
}
```

Maybe I'm not grokking the context, and this is irrelevant as it would only be stuffed into js code... but thought I'd mention just in case.

## `query(callback)` Mystery Functions
The `util.mixin(...)` is a deep object merge fn();

The `.supplant` is taking the spec-json, stringifying it, and interpolating values into it.

**Comment by tomwayson** [comment]
Thx @dbouwman! 

Yeah, using a function in the dataset object doesn't make me feel great. I did a quick glance for any parts of the code that would try to treat the dataset as JSON, but didn't see any (doesn't mean there aren't any). The reason I suggested having it there instead of say an addtional argument passed to `new Cedar()` is just to keep it close to the things it is related to (dataset`url`, `query`, and `mappings`).

**Comment by tomwayson** [comment]
What @benstoltz and I have settled on is adding a new property to Cedar instances, tentatively named `.transform` which would be a function that would take the query results _and_ the dataset as arguments and a modified feature set. Example:
```js
cedar.transform = function(features, dataset) {
  // may only want to do this under certain conditions
  if (!dataset.query.groupByFieldsForStatistics) {
    return features.map(function(feature) 
      // TODO: modify feature attributes here
      return feature;
    });
  } else {
    return features;
  }
}
```

`.update()` would check if `.transform` is defined, and if so, call it.

I would say this should follow the pattern of other properties like `override` where you can pass it to the constructor, or set it later.

Any objections? Any suggestions for the name?

**Comment by tomwayson** [comment]
Resolved in #235
